internal/consensus/state.go

```diff
@@ -1311,24 +1313,33 @@ func (cs *State) enterPrevote(height int64, round int32) {
 	// (so we have more time to try and collect +2/3 prevotes for a single block)
 }
 
+func (cs *State) proposalIsTimely() bool {
+	tp := types.TimingParams{
+		Precision:    cs.state.ConsensusParams.Timing.Precision,
+		MessageDelay: cs.state.ConsensusParams.Timing.MessageDelay,
+	}
+
+	return cs.Proposal.IsTimely(cs.ProposalReceiveTime, tp, cs.state.InitialHeight)
```

Comment:
What is the meaning of `InitialHeight` here? Is it related to the bootstrap/recovery of a node?

Reply:
`InitialHeight` here is the first height _of the chain_. I.e., the genesis height. This means we aren't checking Timely for the genesis height so that chains can start with genesis times in the past.

Reply:
But why would we check genesis proposals? I though the genesis was a block a priori decided. But it seems this is a new feature, isn't it?

Reply:
iiuc, the first consensus state is created from `genesis.json` that includes something like 
```
     "initial_height": "0"
     "genesis_time": "2022-01-07T11:41:10.911589Z"
```
 Then consensus proposes the first block (height 1) with an empty commit. When the first block is created we use the genesis time (from `genesis.json`).
https://github.com/tendermint/tendermint/blob/bd99771b272afbf370e0ca290421673da3c76f14/internal/state/state.go#L268-L272

If we would put the `time.Now()` in this case we would fail in `ValidateBlock` down here:
https://github.com/tendermint/tendermint/blob/bd99771b272afbf370e0ca290421673da3c76f14/internal/state/validation.go#L118-L125

If we relax this check to be `if !block.Time.After(state.LastBlockTime)` then everything seems to be fine and I can bootstrap a chain. I don't know why we need block 1 to have the same timestamp as the genesis time (which is for virtual block 0)

Reply:
> But why would we check genesis proposals? I though the genesis was a block a priori decided. But it seems this is a new feature, isn't it?

The first block is produced through consensus. This is not a change, it's the historical behavior. It's produced with the values from the genesis block.

AFAICT, the point of `genesis_time` is to indicate _at what time the initial block of the block chain should occur_. Validators wait until the genesis time before beginning consensus. I see no strong first principles justification for the block to have the same timestamp as what is written in the `genesis.json` file. That file, from what I can tell, is mostly a social consensus for when the blockchain should begin. Additionally, it will not break existing chains to update the strict equality to be `>=` since this is a relaxation of a previous constraint. 

For these reasons, I'm in favor of having the initial height be >= not strictly =.

Reply:
moved the changes for first block (height 1) to use `time.Now()` to https://github.com/tendermint/tendermint/pull/7541